### PR TITLE
[curl] Accept all supported content encodings by default

### PR DESCRIPTION
**Pull request by candrews**
## Description
By setting CURLOPT_ACCEPT_ENCODING to "" by default, curl will request all
supported content encodings.
Users can still use the "Encoding" protocol option to override this
behavior; setting "Encoding" to "" will cause CURLOPT_ACCEPT_ENCODING
to not be set, resulting in curl not accepting any content encoding
(matching the default behavior before).

## Motivation and Context
I saw https://github.com/xbmc/xbmc/pull/15628 and wanted to improve upon that proposed solution.
This approach is basically what was suggested by @bol-van in the comment at https://github.com/xbmc/xbmc/pull/15628/#issuecomment-468284855

## How Has This Been Tested?
Compiled on Linux then used for an hour or so, mostly playing Youtube videos, looking at my security camera, and looking at pictures from my web server. I didn't notice any problems.

I setup my web server log the value of incoming request's `Accept-Encoding` header. With this PR, Kodi now sends `deflate, gzip, br` as is desired.

## Screenshots (if appropriate):
n/a

## Types of change

- [X] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:

- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [X] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [ ] All new and existing tests passed

### CODE REVIEW

**Comment by pkerling** [comment]
@candrews What do you think about removing the few instances in the code that set a specific requested encoding? I think we could get rid of them as part of this PR.

**Comment by bol-van** [comment]
Pls check.  Something is not good with brotli in kodi

I tried this list in pvr.iptvsimple
works fine :
 http://91.92.66.82/trash/ttv-list/ttv.all.tag.iproxy.m3u|encoding=gzip
kodi hangs, although in wireshark query succeeds : 
 http://91.92.66.82/trash/ttv-list/ttv.all.tag.iproxy.m3u|encoding=br

In wireshark I can see http query but next packet from client is "FIN, ACK".
Looks like after sending headers kodi immediately closes socket
Everything send by the server is ignored then

Curl works good with --compressed parameter and server returns br response

**Comment by candrews** [comment]
> @candrews What do you think about removing the few instances in the code that set a specific requested encoding? I think we could get rid of them as part of this PR.

Sounds good to me - I found two instances (one in Repository, one in ScraperUrl) that set the encoding and have removed each of them in separate commits that I added.

**Comment by bol-van** [comment]
In linux brotli behaves differently but still not good.
Only 295 channels of 1300+ are loaded. File seem to be truncated

**Comment by candrews** [comment]
@bol-van this sounds like a special case of a problem with this particular server - perhaps you should set the url protocol option to only use gzip?

**Comment by bol-van** [comment]
> @bol-van this sounds like a special case of a problem with this particular server - perhaps you should set the url protocol option to only use gzip?

Can you offer any example url that works with br ?

**Comment by candrews** [comment]
I'm staying if this particular server has a problem with br, explicitly request something else.
For example, you said earlier:
http://91.92.66.82/trash/ttv-list/ttv.all.tag.iproxy.m3u|encoding=gzip
Use that and avoid brotli in this case.

Fwiw, I haven't had a problem with any urls I've tried and some have used brotli.

**Comment by bol-van** [comment]
I still think its worth studying why kodi fails with this particular url.
Because chrome, firefox, curl - all download without problems
May be brotli lib is outdated in kodi and thats why it misbehaves

**Comment by bol-van** [comment]
I checked userdata/addon_data/pvr.iptvsimple/iptv.m3u.cache
And this file is not plain. its brotli encoded
I successfully uncompressed it with brotli tools
So kodi *fails* to decompress result from server and tries to parse compressed data

**Comment by candrews** [comment]
Did you build Kodi yourself? Are you using the changes from this PR?

**Comment by bol-van** [comment]
I use 18.0 official build in windows and https://launchpad.net/~team-xbmc/+archive/ubuntu/ppa
for ubuntu 16

**Comment by candrews** [comment]
The official builds don't include brotli support, so that explains the behavior you're seeing :-)

**Comment by bol-van** [comment]
> The official builds don't include brotli support, so that explains the behavior you're seeing :-)

OK, that explains. then by default it should not put "br" to Accept-Encoding.
ubuntu 18.10 kodi 17.6   playlist not downloaded at all with BR

**Comment by candrews** [comment]
This PR will set Accept-Encoding to whatever curl supports. On my system, curl supports brotli, so it gets included. On yours, it doesn't, so it won't be included.

**Comment by pkerling** [comment]
@candrews Did you test the change in `CScraperUrl`? I *guess* it should work fine but would be good to be safe here.

**Review by yol — APPROVED** [review]

**Comment by candrews** [comment]
I'm using Kodi with this patch and haven't had any issues. Can you suggest something I can do to exercise that path? For example, is playing a YouTube video sufficient?

**Comment by pkerling** [comment]
Unfortunately not, sorry. I'd have to look myself where and how this is used. This is also why I'm not completely sure whether it would work or not :-) I guess you'll have to try something with scrapers?

**Comment by wsnipex** [comment]
@MilhouseVH can you include this in your builds please?

**Comment by MilhouseVH** [comment]
@wsnipex will include it in the next nightly, `#0304` on Monday.

**Comment by MilhouseVH** [comment]
I've had negative reports in the forum relating to this PR - not including/reverting the PR resolves issues with seeking "live" streams (YouTube etc.):

https://forum.kodi.tv/showthread.php?tid=298461&pid=2832516#pid2832516
https://forum.kodi.tv/showthread.php?tid=298461&pid=2832710#pid2832710
https://forum.kodi.tv/showthread.php?tid=298461&pid=2832834#pid2832834
https://forum.kodi.tv/showthread.php?tid=298461&pid=2832836#pid2832836

**Comment by bol-van** [comment]
Ideally code should destinguish what are we getting now : live stream or something else.
May be leave all encodings but disable them if byte range specified ?

**Comment by MilhouseVH** [comment]
The easiest test the YouTube problem is to:

1. Play a video from YouTube (anything from "Popular now" should do)
2. Seek forwards a couple of minutes
3. Try to seek in reverse (ie. backwards)

With this PR, every time you seek backwards (ie. `Seek(-7)`, or `StepBack`) you will only ever seek *forwards* by the same amount.

With this PR, seeking forwards works only somewhat - it only ever seeks a small amount of time forward (about 10-20 seconds) even though you may attempt to incrementally seek several several minutes ahead. Jumping forward or backward to a specific timecode such as 5:00 also doesn't work.

This behaviour is reproducible on RPi2 and Generic (x86_64) LibreELEC builds with this PR, and also Ubuntu 17.10 with latest Kodi master ac9206b2c559028bd6472e0c6cda5c88a149e4b7 + 15644.

So in summary, seeking both forwards and backwards within a live stream is broken by this PR.

**Comment by candrews** [comment]
I think what's happening is that Kodi is sending the request with the `Accept-Encoding` header set to the various encodings (which is what this PR does) and the server is responding with content that's not encoded. However, Kodi currently assumes that if it requested encoded content, then the response is encoded - but that's not correct in this case.

My PR at https://github.com/xbmc/xbmc/pull/15643 fixes that problem by looking at the response (not the request!) to determine if the response is encoded.

I won't be able to test this myself until later tonight, but if someone else wants to test using both this PR and https://github.com/xbmc/xbmc/pull/15643 to see if the live stream problem is solved, I'd greatly appreciate it :)

**Comment by MilhouseVH** [comment]
@candrews spot on, with #15643 now merged this PR seems to be working perfectly with YouTube content. I'll continue to include this PR in nightlies for feedback from those that reported issues, but it's looking much better now. Thanks!

**Comment by MilhouseVH** [comment]
Both negative reports are now positive with the latest nightly build.

**Review by wsnipex — APPROVED** [review]

**Comment by candrews** [comment]
Can this now be merged?

Thanks again!

**Comment by Rechi** [comment]
If now all supported encodings are accepted by default, is there any use case where one would want to limit the accepted encodings?

https://github.com/xbmc/xbmc/blob/05fff94ec9064c3ff914a10c0857408a79bb58dd/xbmc/filesystem/CurlFile.cpp#L797-L798

**Comment by bol-van** [comment]
Who knows. There may be misbehaving servers. I see no reason in removing already present functionality such as appending "|encoding=" to url.  It can be used as a quick fix

**Comment by wsnipex** [comment]
I'd keep it as well, at least for v19

**Comment by candrews** [comment]
Shall this be merged then?

**Review by yol — APPROVED** [review]

**Comment by pkerling** [comment]
Didn't seem like there were any objections, thanks for contributing :tada:
